server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -391,29 +431,131 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+    private static Map<String, Role> roleNameToPossibleRoles;
+
+    public static void setPossibleRoles(final Set<Role> possibleRoles) {
+        final Map<String, Role> roleNameToPossibleRoles =
+                possibleRoles.stream().collect(Collectors.toUnmodifiableMap(Role::roleName, Function.identity()));
+        // collect the abbreviation names into a map to ensure that there are not any duplicate abbreviations
+        final Map<String, Role> roleNameAbbreviationToPossibleRoles = roleNameToPossibleRoles.values()
+                .stream()
+                .collect(Collectors.toUnmodifiableMap(Role::roleNameAbbreviation, Function.identity()));
+        assert roleNameToPossibleRoles.size() == roleNameAbbreviationToPossibleRoles.size() :
+                "roles by name [" + roleNameToPossibleRoles + "], roles by name abbreviation [" + roleNameAbbreviationToPossibleRoles + "]";
+        DiscoveryNode.roleNameToPossibleRoles = roleNameToPossibleRoles;
+    }
+
+    public abstract static class Role {
+
+        private final String roleName;
+
+        public String roleName() {
+            return roleName;
+        }
+
+        private final String roleNameAbbreviation;
+
+        public String roleNameAbbreviation() {
+            return roleNameAbbreviation;
+        }
+
+        protected Role(final String roleName, final String roleNameAbbreviation) {
+            this.roleName = roleName;
+            this.roleNameAbbreviation = roleNameAbbreviation;
+        }
+
+        protected abstract Setting<Boolean> roleSetting();
+
+        @Override
+        public String toString() {
+            return "Role{" +
+                    "roleName='" + roleName + '\'' +
+                    ", roleNameAbbreviation='" + roleNameAbbreviation + '\'' +
+                    '}';
+        }
+
+    }
+
+    public static class MasterRole extends Role {
+
+        public static final MasterRole INSTANCE = new MasterRole();
+
+        private MasterRole() {
+            super("master", "m");
+        }
+
+        @Override
+        protected Setting<Boolean> roleSetting() {
+            return Node.NODE_MASTER_SETTING;
+        }
+
+    }
+
+    public static class DataRole extends Role {
+
+        public static final DataRole INSTANCE = new DataRole();
+
+        private DataRole() {
+            super("data", "d");
+        }
+
+        @Override
+        protected Setting<Boolean> roleSetting() {
+            return Node.NODE_DATA_SETTING;
+        }
+
+    }
+
+    public static class IngestRole extends Role {
+
+        public static final IngestRole INSTANCE = new IngestRole();
+
+        private IngestRole() {
+            super("ingest", "i");
+        }
+
+        @Override
+        protected Setting<Boolean> roleSetting() {
+            return Node.NODE_INGEST_SETTING;
+        }
+
+    }
+
+    private static class UnknownRole extends Role {
```

Comment:
Can you explain why this is needed? I can't tell the purpose, and based solely on the name it seems like some type of leniency.

Reply:
If 7.4 adds a new role a 7.3 node won’t know it, so 7.3 needs to accept roles that are unknown to it. Similarly if 8.0 removes a role 7.8 knows about, 8.0 needs to account for receiving a role that is unknown to it.